python/pyspark/sql/types.py

```diff
@@ -368,8 +393,47 @@ def __init__(self, fields):
         >>> struct1 == struct2
         False
         """
-        assert all(isinstance(f, DataType) for f in fields), "fields should be a list of DataType"
-        self.fields = fields
+        if not fields:
+            self.fields = []
+        else:
+            self.fields = fields
+            assert all(isinstance(f, StructField) for f in fields),\
+                "fields should be a list of StructField"
+
+    def add(self, name_or_struct_field, data_type=None, nullable=True, metadata=None):
```

Comment:
can we change name_or_struct_field to just "name", and just explain we support passing in a struct type in the docstring?